text/1004-transactions-api.rst

```diff
@@ -403,107 +601,197 @@ where we catch network error before sending a request.
+
+    async for tx in db.retry():
+      async with tx:
+        tx = tx.with_metadata(
+            uri=request.uri,
+            username=request.session.get('username'),
+        )
+        tx.query("...")
+
 
-But there are couple of issues:
+More Read-Only Options
+----------------------
 
-1. Repeating read-only queries is always safe, but we don't know which
-   ones are readonly. We tackle this below.
-2. Repeating non-readonly queries can be dangerous: they may be applied
-   twice.
+Replica config may be specified when configuring a read-only connection:
 
-There are couple of ideas to differentiate read-only and mutable
-queries:
+.. code-block:: python
 
-1. We can expose "read-only" flag in ``Prepare`` (and it's always safe
-   to retry before ``Execute`` happens)
-2. Enable connection mode that forbids mutating queries outside of
-   transaction and retry all (read-only) queries
-3. Do (2) and have some method to override it on per-method-call basis
-4. Or vice versa add ``pool.read_only_query()`` method
-5. Add ``pool.read_only().query()``
-6. Add ``pool.with_options({readonly: true}).query()``
+   def read_only(self, primary=False, max_replica_lag=10): ...
 
-Note: methods (2-6) are also helpful for working with primary/replica
-installations. But probably only last two would allow full power, as they allow
-``pool.read_only(primary=true)`` (i.e. in case you need read-only transaction
-that can't go to a replica).
+Perhaps this should be encapsulated into replica options:
 
-This issue can be solved by a later RFC.
+.. code-block:: python
+
+   def read_only(self, primary=False, replicas: ReplicaOptions): ...
 
 
 Learning Curve
 ==============
 
-This complicates learning curve, but:
+The ``retry`` method complicates learning curve curve, but:
 
 1. This is already a problem in Postgres, there a lot of people who
-   ignore the issue for pet projects and a lot of startups having "normal"
-   rate or 500 errors at any point of time, while it's preventable.
+   ignore the issue for pet projects and a lot of startups having
+   "normal" rate or 500 errors at any point of time, while it's
+   preventable.
 2. Repeating transactions would make less incentive to keep transactions
-   open for a long time (e.g. while accessing slow network resources like
-   external API), which is a problem of itself.
+   open for a long time (e.g. while accessing slow network resources
+   like external API), which is a problem of itself.
 3. Even if we never have failed concurrent updates we would want
    seamless reconnect on connection failures (i.e. server restart,
    primary/replica change, etc.)
 4. To make learning curves shorter I think we should intentionally
-   inject failures. This is needed so that users quickly find out that side
-   effects of their transactions are in effect several times.
+   inject failures. This is needed so that users quickly find out that
+   side effects of their transactions are in effect several times.
```

Comment:
"4. In the future we will consider adding a connection setting to inject failures into transactions to help users discover incorrect transactional code."

@tailhook @elprans I think we shouldn't do this before 1.0 as I'm not sure how it might impact benchmarks, production code, and the overall DX of early adopters. I think that our API is already going to be structured in the right way, and adding failure injection is entirely optional at this point.

Reply:
I also think that proper API design, documentation and tutorials would be sufficient to guide users toward the correct set of idioms.  Intentional failure injection might be an optional feature.

Reply:
This is where we constantly disagree. Documentation and tutorials are not enough for the same reason for which we use unit tests to test code. As we now basically know the distinction between workstation installations (created by `edgedb server init`) and production (roughly -- everything else), it's possible to have server itself without failure injection by default, but `server init` configuring it in a way that enables this by default. Also that section describes a way of failure injection that doesn't impact benchmarks in any significant way.

But anyway, this RFC is already hanging for too long, so I drop the "failure injection" section in favor of later RFC.